votingpool/withdrawal.go

```diff
@@ -0,0 +1,939 @@
+	}
+
+	// Re-add the last item from removedInputs, which is the last popped input.
+	tx.addInput(removedInputs[len(removedInputs)-1])
+	removedInputs = removedInputs[:len(removedInputs)-1]
+	return removedInputs, removedOutput, nil
+}
+
+func newWithdrawal(roundID uint32, requests []OutputRequest, inputs []Credit,
+	changeStart ChangeAddress) *withdrawal {
+	outputs := make(map[OutBailmentID]*WithdrawalOutput, len(requests))
+	for _, request := range requests {
+		outputs[request.outBailmentID()] = &WithdrawalOutput{request: request}
+	}
+	status := &WithdrawalStatus{
+		outputs:        outputs,
+		nextChangeAddr: changeStart,
+	}
+	return &withdrawal{
+		roundID:         roundID,
+		current:         newWithdrawalTx(),
+		pendingRequests: requests,
+		eligibleInputs:  inputs,
+		status:          status,
+		newWithdrawalTx: newWithdrawalTx,
+	}
+}
+
+// StartWithdrawal uses a fully deterministic algorithm to construct
+// transactions fulfilling as many of the given output requests as possible.
+// It returns a WithdrawalStatus containing the outpoints fulfilling the
+// requested outputs and a map of normalized transaction IDs (ntxid) to
+// signature lists (one for every private key available to this wallet) for each
+// of those transaction's inputs. More details about the actual algorithm can be
+// found at http://opentransactions.org/wiki/index.php/Startwithdrawal
+func (p *Pool) StartWithdrawal(roundID uint32, requests []OutputRequest,
+	startAddress WithdrawalAddress, lastSeriesID uint32, changeStart ChangeAddress,
+	txStore *txstore.Store, chainHeight int32, dustThreshold btcutil.Amount) (
+	*WithdrawalStatus, error) {
+
+	eligible, err := p.getEligibleInputs(txStore, startAddress, lastSeriesID, dustThreshold,
+		chainHeight, eligibleInputMinConfirmations)
+	if err != nil {
+		return nil, err
+	}
+
+	w := newWithdrawal(roundID, requests, eligible, changeStart)
+	if err := w.fulfillRequests(); err != nil {
+		return nil, err
+	}
+	w.status.sigs, err = getRawSigs(w.transactions)
+	if err != nil {
+		return nil, err
+	}
+
+	return w.status, nil
+}
+
+// popRequest removes and returns the first request from the stack of pending
+// requests.
+func (w *withdrawal) popRequest() OutputRequest {
+	request := w.pendingRequests[0]
+	w.pendingRequests = w.pendingRequests[1:]
+	return request
+}
+
+// pushRequest adds a new request to the top of the stack of pending requests.
+func (w *withdrawal) pushRequest(request OutputRequest) {
+	w.pendingRequests = append([]OutputRequest{request}, w.pendingRequests...)
+}
+
+// popInput removes and returns the first input from the stack of eligible
+// inputs.
+func (w *withdrawal) popInput() Credit {
+	input := w.eligibleInputs[0]
+	w.eligibleInputs = w.eligibleInputs[1:]
+	return input
+}
+
+// pushInput adds a new input to the top of the stack of eligible inputs.
+func (w *withdrawal) pushInput(input Credit) {
+	w.eligibleInputs = append([]Credit{input}, w.eligibleInputs...)
```

Comment:
I wonder if you should reverse the stack semantics here.  This will cause a lot of extra work and garbage because it will create a a new single element slice and append the rest of the stack to it, thereby potentially requiring several grows.  Each time a slice has to grow, it has to copy all elements.

So like:

```Go
w.eligibleInputs = append(w.eligibleInputs, input)
```

And then for `popInput`:

```Go
input := w.eligibleInputs[len(w.eligibleInputs)-1]
w.eligibleInputs = w.eligibleInputs[:len(w.eligibleInputs)-2]
return input
```